rfcs/greedy-stageless.md

```diff
@@ -0,0 +1,233 @@
+
+The simple solution to this problem is to separate whether a system should run from whether the run criteria should be checked again. Run Criteria becomes a boolean and a new type of system called Looping Criteria returns a boolean with whether the run criteria should be checked again.
+
+### Implementation of Looping Criteria
+
+Systems with a looping criteria are run in a loop until the looping criteria returns false. Systems with looping criteria are considered a blocking dependency to systems that depend on them until the looping criteria returns false.
+
+A problem with looping criteria that will need to be solved is that they will need to work across hard sync points. i.e. If a looping criteria is applied to a system set. The looping criteria can apply to a sub graph that might require multiple hard syncs before reevaluating the run criteria. Looping criteria will need to keep track of whether the systems dependent on it have run for the current loop before evaluating.
+
+It might be possible to reuse the logic for the system graph and consider looping criteria to have a `after` dependency on all systems that have a dependency on the looping criteria.
+
+### When should run criteria be evaluated?
+
+Currently bevy evaluates all run criteria once before any systems are run and again after all parallel systems run. The proposed loop does not have an equivalent to this as parallel systems are split up between hard sync points. I think there are a couple ways of doing this.
+
+- Before step 1
+- Before step 1 and before step 2
+
+Not sure which would be preferred or if there are any options closer to the current model.
+
+### Iteration count statistic
+
+A statistic should be added that says what iteration of the run systems loop it was executed on. Knowning in which loop systems are run will tell the user which systems do not block each other and could help with debugging other scheduling problems.
+
+### Bye-Bye Stages
+
+Stages will be removed as part of this RFC. Not sure on the exact plan yet, but I think it should be straightforward.
+
+Current stages will probably be replaced by system sets. But chould also be redone with more mind payed to the data flow.
+
+## Drawbacks
+
+- This makes a bevy tick a lot less linear and harder to reason about. There will probably need to be good visualization tools to help deal with this.
+- Mapping the current stages onto the new method may get confusing. If a
+  system is in PREUPDATE and another system is in POSTUPDATE what is the relationship between the two? Are they related at all? Are they implicitly depending on something happending in UPDATE? It will take some careful thought to correctly map the current stages and systems to this new api.
```

Comment:
A no-brainer solution would be:

```
PREUPDATE -> PREUPDATE(commands) -> UPDATE -> UPDATE(commands) -> POSTUPDATE
```

Maybe not the best solution, but the behavior of the three systems should not change.